spices/0002-uri-rewriting-for-mirroring.adoc

```diff
@@ -0,0 +1,319 @@
+
+[source,pkl]
+----
+@Since { version = "0.26.0" }
+class UriRewrite {
+  /// What to match in the given URI.
+  replace: String|Regex
+
+  /// What to replace the _all_ matches with.
+  ///
+  /// When `replace` is a `String`, `with` is interpreted literally.
+  /// When `replace` is a `Regex`, `with` is interpreted as a replacement pattern,
+  /// which may contain references to capture groups (`$0`, `$1`, etc).
+  with: String
+}
+
+/// Rewrite rules for URIs used in module keys and resource reads.
+///
+/// For any URI, the first of the longest-matching [UriRewrite]s is applied, before applying the secutiry manager.
+@Since { version = "0.26.0" }
+uriRewrites: Listing<UriRewrite>
+----
+
+=== Common CLI flags
+
+These flags are _common_, because this setting applies to different commands equally.
+The proposed `analyze` command (see link:../spices/0001-import-graph-analyzer.pkl[SPICE 1: Import Graph Analyzer API]), for example, applies these rewrites in the resolution of all URIs analyzed and reported.
+
+This proposal adds two flags to the CLI:
+  * `--uri-rewrite <regex> <replacement>`, which adds a URI rewrite rule; and
+  * `--no-uri-rewrites`, which discards all URI rewrite rules defined thus far.
+
+If a `pkl:Project` file is found, this is read first.
+To invoke Pkl without rewrite rules defined therein, `--no-uri-rewrites` can be used, followed by zero-or-more `--uri-rewrite` arguments.
+
+=== Rewrite behavior
+
+The proposed design is to match any URI being resolved in its entirety to all the provided rewrites' `replace` specification.
+The first, longest-matching rewrite rule is applied.
+In other words:
+ - when multiple rules match, a rule that matches a longer substring of the URI is preferred over one matching a shorter substring; and
+ - when multiple rules match the same substring length, the first rule is chosen (in definition order).
+
+The intended rewrite behavior, expressed as Pkl, is
+[source, pkl]
+----
+class Rewrite {
+  rewrites: List<UriRewrite>
+  uri: Uri
+
+  matchesByScore = rewrites.groupBy((rewrite) ->
+    if (!uri.contains(rewrite.replace))
+      -1
+    else if (rewrite.replace is Regex)
+      let (match = rewrite.replace.findMatchesIn(uri).first)
+        match.end - match.start
+    else
+      rewrite.replace.length
+  )
+
+  maxScore = matchesByScore.keys.max
+
+  rewritten: Uri = if (maxScore == -1) uri else
+    let (rewrite = matchesByScore[maxScore].first)
+      uri.replaceFirst(rewrite.replace, rewrite.with)
+}
+
+function rewrite(inputUri: Uri, uriRewrites: List<UriRewrite>): Uri = new Rewrite {
+  rewrites = uriRewrites
+  uri = inputUri
+}.rewritten
+----
+
+=== Implementation
+
+The proposed design is to implement this behaviour in `IoUtils::resolve`.
+`IoUtils::resolve` is used throughout to resolve URIs, regardless of whether a URI is used in an `import`, `import*`, `extends`, `amends`, `read`, or `read*`.
```

Comment:
This should probably use `org.pkl.core.http.RequestRewritingClient` instead, because `IoUtils.resolve` doesn't affect how packages get loaded.

We can definitely use `org.pkl.core.http.RequestRewritingClient` if we only permit `http`/`https` rewrites. It'd be more problematic if we allow other types of rewrites though.

Reply:
^^ This goes back to whether we want to restrict rewriting to HTTP(S) and ignore custom resource readers.